crates/full-node/sov-ethereum/src/handlers/subscribe/service.rs

```diff
@@ -0,0 +1,172 @@
+use super::watermark::Watermark;
+use crate::Ethereum;
+use alloy_consensus::Sealed;
+use alloy_rpc_types::Filter;
+use alloy_rpc_types::Header;
+use jsonrpsee::DisconnectError;
+use jsonrpsee::SubscriptionMessage;
+use jsonrpsee::SubscriptionSink;
+use serde::Serialize;
+use sov_address::{EthereumAddress, FromVmAddress};
+pub use sov_evm::EthereumAuthenticator;
+use sov_evm::Evm;
+use sov_evm::MaybeSealedBlock;
+use sov_evm::Receipt;
+use sov_modules_api::capabilities::HasKernel;
+use sov_modules_api::ApiStateAccessor;
+use sov_modules_api::Spec;
+use sov_sequencer::Sequencer;
+use std::fmt::Debug;
+use std::sync::Arc;
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Block does not exist")]
+    BlockDoesNotExist,
+    #[error("Receipt does not exist")]
+    ReceiptDoesNotExist,
+    #[error(transparent)]
+    Disconnect(#[from] DisconnectError),
+}
+
+pub struct Streamer<S, Seq>
+where
+    S: Spec,
+    Seq: Sequencer<Spec = S>,
+{
+    sink: SubscriptionSink,
+    ethereum: Arc<Ethereum<S, Seq>>,
+    evm: Evm<S>,
+}
+
+impl<S, Seq> Streamer<S, Seq>
+where
+    S: Spec,
+    Seq: Sequencer<Spec = S>,
+    S::Address: FromVmAddress<EthereumAddress>,
+    Seq::Rt: HasKernel<S> + EthereumAuthenticator<S> + Default + Send + Sync + 'static,
+{
+    pub fn new(sink: SubscriptionSink, ethereum: Arc<Ethereum<S, Seq>>) -> Self {
+        Self {
+            sink,
+            ethereum,
+            evm: Default::default(),
+        }
+    }
+
+    /// Stream new logs matching the provided filter to the subscriber.
+    pub async fn logs(&self, filter: Box<Filter>) -> Result<(), Error> {
+        let mut state = self.ethereum.api_state_accessor();
+        let pending_block = self.evm.pending_block(&mut state);
+        let mut tx_watermark = Watermark::new(pending_block.transactions.end);
+
+        // Fetch the initial block. If it's stale, it will be replaced below.
+        let mut block = self.get_block(pending_block.header.number - 1, &mut state)?;
+
+        let mut state_updates = self.ethereum.sequencer.api_state().checkpoint_receiver();
+        while state_updates.changed().await.is_ok() {
+            let mut state = self.ethereum.api_state_accessor();
+            let pending_block = self.evm.pending_block(&mut state);
+
+            for tx_idx in tx_watermark.advance(pending_block.transactions.end) {
```

Comment:
It looks like if `pending_block.transactions.end = 7`, we will try to iterate over index 7 here. However, the `ending_block.transactions` range is exclusive, so it appears we’re attempting to access a transaction index that doesn’t exist.

Reply:
Fixed